chicago/permit_cleaning.py

```diff
@@ -31,33 +31,103 @@
+        print(
+            f"Invalid end_date format: '{end_date_str}'. Expected YYYY-MM-DD."
+        )
+        sys.exit(1)
+
+    if end_date < start_date:
+        print("Error: end_date must be later than or equal to start_date.")
+        sys.exit(1)
+
+    deduplicate = deduplicate.lower() == "true"
+
+    return start_date_str, end_date_str, deduplicate
 
 
-# Connect to Athena and download existing 14-digit PINs in Chicago
-def pull_existing_pins_from_athena(cursor):
-    print("Pulling PINs from Athena")
+def year_from_date_string(date_str: str) -> str:
+    """Parse a date string in YYYY-MM-DD format and return a string representing
+    the year of the date"""
+
+    return str(datetime.strptime(date_str, "%Y-%m-%d").year)
+
+
+def get_pin_cache_filename(start_date: str, end_date: str) -> str:
+    """Given start and end dates, return the name of a file that we can use to
+    cache distinct PINs between the years represented by the two dates"""
+
+    # Assume that dates are already validated for YYYY-MM-DD format
+    start_year = year_from_date_string(start_date)
+    end_year = year_from_date_string(end_date)
+
+    return f"chicago_pin_universe-{start_year}-{end_year}.csv"
+
+
+def pull_existing_pins_from_athena(
+    cursor: Cursor, start_date: str, end_date: str
+) -> pd.DataFrame:
+    """Connect to Athena and download all PINs in Chicago between the given
+    start and end dates"""
+
+    # Assume that dates are already validated for YYYY-MM-DD format
+    start_year = year_from_date_string(start_date)
+    end_year = year_from_date_string(end_date)
+
     SQL_QUERY = """
-        SELECT
+        SELECT DISTINCT
             CAST(pin AS varchar) AS pin,
             CAST(pin10 AS varchar) AS pin10
         FROM default.vw_pin_universe
-        WHERE triad_name = 'City' AND year = '2023'
+        WHERE triad_name = 'City'
+            AND year BETWEEN %(start_year)s AND %(end_year)s
     """
-    cursor.execute(SQL_QUERY)
+    cursor.execute(SQL_QUERY, {"start_year": start_year, "end_year": end_year})
     chicago_pin_universe = as_pandas(cursor)
-    chicago_pin_universe.to_csv("chicago_pin_universe.csv", index=False)
+    pin_cache_filename = get_pin_cache_filename(start_date, end_date)
+    chicago_pin_universe.to_csv(pin_cache_filename, index=False)
 
     return chicago_pin_universe
 
 
-def download_permits(start_date, end_date):
+def download_permits(start_date: str, end_date: str) -> pd.DataFrame:
+    """Download permits from the Chicago open data portal in the dataframe
+    with issue dates between `start_date` and `end_date`"""
     params = {
+        # Assume we've already validated the start and end date strings for
+        # YYYY-MM-DD format
         "$where": f"issue_date between '{start_date}' and '{end_date}'",
         "$order": "issue_date DESC",
-        "$limit": 1000000,  # Artificial limit to override the default
+        "$limit": 10000000,  # Artificial limit to override the default
```

Comment:
There are [~800k rows in the source data](https://data.cityofchicago.org/Buildings/Building-Permits/ydr8-5enu/about_data), so I bumped this limit in case we want to query a larger proportion of them at some point in the future.